package/MDAnalysis/analysis/base.py

```diff
@@ -177,13 +170,16 @@ def run(self, start=None, stop=None, step=None, verbose=None):
         self._setup_frames(self._trajectory, start, stop, step)
         logger.info("Starting preparation")
         self._prepare()
-        for i, ts in enumerate(
-                self._trajectory[self.start:self.stop:self.step]):
+        verbose = getattr(self, '_verbose', False)
+        n_frames = len(range(*self._trajectory.check_slice_indices(
```

Comment:
Rather than calculating the n_frames, I thought that a trajectory slice knew it's length?  Ie `len(u.trajectory[::2])` should work, so maybe tqdm can infer the length of the traj slice

Reply:
That would be true if there wasn't an `enumerate` around the traj slice, and we need it here ! But I just realised I can remove the `n_frames` for hbond_analysis.py, remove the `enumerate` in density.py, helanal.py and remove both `n_frames` and `enumerate` in pca.py. Thanks !

Reply:
You should still use len on the sliced trajectory to get n_frames. It would make the code more readable and avoid potential issues we had mixing slices and range.

Reply:
Makes sense ! I also just realised I can put the ProgressBar inside enumerate instead of the other way around, which makes it even simpler.